### ISSUE TITLE
Reauth not occuring?

### ISSUE THREAD

**Issue by sb10** [gophercloud/gophercloud#1324]
I've been using AllowReauth with success and never thought about how and under what circumstances it works, but I hit a situation where I started getting "Authentication failed" and it never recovered.

I set up like:

```
opts, err := openstack.AuthOptionsFromEnv()
opts.AllowReauth = true
provider, err := openstack.AuthenticatedClient(opts)
computeClient, err = openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
	Region: os.Getenv("OS_REGION_NAME"),
})
```

This then works fine when I use the computeClient to do things like create new servers over a period of days. Normally I don't get auth issues for any other usage either.

But on one occasion I got stuck here:

```
        pager := flavors.ListDetail(computeClient, flavors.ListOpts{})
	return pager.EachPage(func(page pagination.Page) (bool, error) {
		flavorList, err := flavors.ExtractFlavors(page)
		if err != nil {
			return false, err
		}

		for _, f := range flavorList {
			[...]
		}
		return true, nil
	})
```

This always returned err "Authentication failed" every time I called it. Since that fails it's possible I'm not attempting any other usage of computeClient.

Starting a new process with the exact same executable that calls the above code, with the same OS_* environment variables, from the same host, at the same time as that process continued to get "Authentication failed", showed no authentication issues.

In the above context, should Reauth be occurring? Or am I supposed to string match the err for "Authentication failed" and do a manual reauth? How?

**Comment by mape90** [comment]
I have seen this same problem and it actually happens really often. I was able to identify the error and it seems to be because commit (3bb4676e61b8eda86b9d50ab2906708b78017115)

There the change isn't thread safe and in some cases other thread can fetch the nil function and then save it after its Request. After this no reauthentications will be made.

```
        reauthFunc := client.ReauthFunc
        client.ReauthFunc = nil
	resp, err = client.Request(method, url, options)
	client.ReauthFunc = reauthFunc
```

For solving this could it be possible to pass explicitly do we want to reauth or not? For example add parameter to `RequestOpts` to force reauth off. So that we could just copy the opts struct and set there the reauth off when we make the recursive `Request` call.

So we could do something like deep copy the options struct and set there the `optionsCopy.ReauthForceOff == true` and then change the reauth check from `client.ReauthFunc != nil ` to `( client.ReauthFunc != nil &&  options.ReauthForceOff == false )` and of cource add that `ReauthForceOff` to the `RequestOpts`

**Comment by sb10** [comment]
Yes, that sounds like a good solution. Can you create a PR for that? Or should I? Someone else?

**Comment by sb10** [comment]
Actually, a simpler solution might be:

```
reauthFunc := client.ReauthFunc
client.ReauthFunc = nil
resp, err = client.Request(method, url, options)
if reauthFunc != nil {
    client.ReauthFunc = reauthFunc
}
```

There's no need to set it back to nil if it was really supposed to be nil, but if we wanted Reauth and hit this edge case where we picked up nil, we don't remove reauth.

**Comment by jtopjian** [comment]
This issue is very similar to #1244. The same patch seems to be causing both problems.

The proposed solution of checking if `reauthFunc != nil` won't work as-is - there needs to be some sort of locking or else we're back to square one: there's a chance that two threads will hit that area of code at the same time.

With that in mind, after reviewing the original issue described in #1182, I'm considering reverting #1192. #1192 describes the following situation:

An OpenStack service is always responding with a 401 regardless of if reauthentication worked.

IMO, for this situation to occur, this is a site-specific / environmental condition and not something that Gophercloud can (or should) handle. If a service is constantly throwing 401s, even if the client can successfully reauthenticate, that points to a problem with the service itself. A custom reauthentication function can be used to help with this kind of situation as described in the [FAQ](https://github.com/gophercloud/gophercloud/blob/master/docs/FAQ.md#implementing-default-logging-and-re-authentication-attempts).

If this is not an appropriate solution, that means the current locking happening in the provider client is not sufficient and reauthentication support needs redesigned since we're constantly playing cat and mouse here.

/cc @garbage-collector

**Comment by jtopjian** [comment]
This should be fixed in #1327
